core/sql/optimizer/SynthType.cpp

```diff
@@ -5207,6 +5207,18 @@ const NAType *Translate::synthesizeType()
          err4106arg = SQLCHARSETSTRING_UTF8;
        break;
 
+     case GBK_TO_UTF8:
+       if (translateSource->getCharSet() == CharInfo::GBK || translateSource->getCharSet() == CharInfo::UnknownCharSet )
+         charsetTarget = CharInfo::UTF8;
+       else
+       {
+            if( CmpCommon::getDefaultString(HIVE_FILE_CHARSET) ==  CmpCommon::getDefaultString(HIVE_DEFAULT_CHARSET) )
+              err4106arg = SQLCHARSETCODE_GB2312;
+            else
+             charsetTarget = CharInfo::UTF8;
```

Comment:
yes, I will add comments. 
Related to what Hans suggested below, this logic need to change. 
It is the basic for this whole task. I will change the logic and add comments